vscode/src/extension.ts

```diff
@@ -1,29 +1,111 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-import { getCompiler, loadWasmModule } from "qsharp";
+import { ILanguageService, getLanguageService, loadWasmModule } from "qsharp";
 import * as vscode from "vscode";
+import { createCompletionItemProvider } from "./completion.js";
+import { createDefinitionProvider } from "./definition.js";
 import { startCheckingQSharp } from "./diagnostics.js";
+import { createHoverProvider } from "./hover.js";
 import { registerQSharpNotebookHandlers } from "./notebook.js";
 
 export async function activate(context: vscode.ExtensionContext) {
+  initializeLogger();
+
+  const languageService = await loadLanguageService(context.extensionUri);
+
+  context.subscriptions.push(
+    ...registerDocumentUpdateHandlers(languageService)
+  );
+
+  context.subscriptions.push(...registerQSharpNotebookHandlers());
+
+  context.subscriptions.push(startCheckingQSharp(languageService));
+
+  // completions
+  context.subscriptions.push(
+    vscode.languages.registerCompletionItemProvider(
+      "qsharp",
+      createCompletionItemProvider(languageService),
+      "."
+    )
+  );
+
+  // hover
+  context.subscriptions.push(
+    vscode.languages.registerHoverProvider(
+      "qsharp",
+      createHoverProvider(languageService)
+    )
+  );
+
+  // go to def
+  context.subscriptions.push(
+    vscode.languages.registerDefinitionProvider(
+      "qsharp",
+      createDefinitionProvider(languageService)
+    )
+  );
+}
+
+function initializeLogger() {
+  const output = vscode.window.createOutputChannel("Q#", { log: true });
+
+  // Override the global logger with functions that write to the output channel
+  global.qscLog.error = output.error;
+  global.qscLog.warn = output.warn;
+  global.qscLog.info = output.info;
+  global.qscLog.debug = output.debug;
+  global.qscLog.trace = output.trace;
+
+  global.qscLog.debug("Q# extension activated.");
```

Comment:
Very minor nit: But I'd consider this an `info` level event. It only occurs once and is useful output.

Reply:
Agreed